### PR TITLE
Use QuantizedPacked for kernel

### PR DESCRIPTION
**Pull request by primenumber**
Introduce QUANTIZED_PACKED_KERNEL type for quantized packed kernel data.

## Motivation and Context

In current implementation of master, quantized packed kernel data type is T_UINT.
T_UINT is easily confused with other data types.
This PR use QuantizedPacked in order to reduce confusing types.

## Description

Input for FPGA has volatile qualifier.
But kernel for FPGA don't have to have volatile qualifier.
So, I introduced QUANTIZED_PACKED_KERNEL type alias.
For convenience, operator^ is generalized for cv qualifier.

C++ version is bumped up to C++14(to infer function return type and use std::remove_cv_t).

Default qword type is changed from Int32 to Uint32 (Because it is only used for QUANTIZED_PACKED, it should be uint32_t).

## How has this been tested?

```shell
$ PYTHONPATH=python/dlk python python/dlk/scripts/generate_project.py -i examples/object_detection/fyolo_quantize_4_v4/minimal_graph_with_shape.pb -o tmp/ -p object_detection -hq -ts
$ cd tmp/object_detection
$ make -j8 lm_x86
$ make clean
$ make -j8 lm_arm
$ make clean
$ make -j8 lm_aarch64
$ make clean
$ make -j8 lm_fpga
```
And run test of lm_* for x86, arm, FPGA (I don't have aarch64 environment).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
@n-nez @antonionevado Please review 🙇

File: dlk/python/dlk/core/data_types.py [hunk 1]
```diff
@@ -80,6 +82,16 @@ def nptype(cls):
         return np.int8
 
 
+class PackedInt8(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: We don't use this, maybe we can deleted it for now?

File: dlk/python/dlk/core/data_types.py [hunk 2]
```diff
@@ -90,6 +102,16 @@ def nptype(cls):
         return np.uint8
 
 
+class PackedUint8(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: Not used?

File: dlk/python/dlk/core/data_types.py [hunk 3]
```diff
@@ -100,6 +122,16 @@ def nptype(cls):
         return np.int16
 
 
+class PackedInt16(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: Not used.

File: dlk/python/dlk/core/data_types.py [hunk 4]
```diff
@@ -110,6 +142,16 @@ def nptype(cls):
         return np.uint8
 
 
+class PackedUint16(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: Not used.

File: dlk/python/dlk/core/data_types.py [hunk 5]
```diff
@@ -120,6 +162,16 @@ def nptype(cls):
         return np.int32
 
 
+class PackedInt32(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: Not used.

File: dlk/python/dlk/core/data_types.py [hunk 6]
```diff
@@ -140,6 +202,16 @@ def nptype(cls):
         return np.int64
 
 
+class PackedInt64(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: More likely to be used soon in 64 bit architectures, so maybe worth keep it.

File: dlk/python/dlk/templates/manual/consts/input.tpl.cpp [hunk 7]
```diff
@@ -25,13 +25,13 @@ limitations under the License.
 {% if node.transposed_data %}
 
 #if defined(RUN_ON_FPGA)
-{{ node.dtype.cpptype() }} {{ node.name }}[] = {
+Base<{{ node.dtype.cpptype() }}>::type {{ node.name }}_raw[] = {
```
Comment (n-nez) [inline dlk/python/dlk/templates/manual/consts/input.tpl.cpp]: How about putting `static` here, to hide implementation? :thinking:

File: dlk/python/dlk/core/data_types.py [hunk 8]
```diff
@@ -150,6 +222,16 @@ def nptype(cls):
         return np.uint64
 
 
+class PackedUint64(Primitive, int):
```
Comment (antonionevado) [inline dlk/python/dlk/core/data_types.py]: Similar case that PackedInt64

File: dlk/python/dlk/core/data_types.py [hunk 9]
```diff
@@ -150,6 +222,16 @@ def nptype(cls):
...
+    @classmethod
+    def cpptype(cls):
+        return 'QuantizedPacked<uint64_t>'
+
+    @classmethod
+    def nptype(cls):
+        return np.uint64
+
+
```
Comment (n-nez) [inline dlk/python/dlk/core/data_types.py]: would prefer to not have types which we do not use

it is easy to add missed ones when we start using them and keeping unused code will increase maintenance effort / mental load while reading.

File: dlk/python/dlk/templates/include/global.tpl.h [hunk 10]
```diff
@@ -35,15 +36,17 @@ template <typename pack_type>
...
  public:
   using T = pack_type;
+  using base_t = std::remove_cv_t<T>;
   static constexpr std::size_t BitCount = sizeof(pack_type) * CHAR_BIT;
   QuantizedPacked() = default;
   explicit QuantizedPacked(const T val) : val(val) {}
-  explicit operator T() const { return val; }
-  QuantizedPacked<T>& operator|=(const QuantizedPacked<T>& that) {
+  explicit operator base_t() const { return val; }
+  template <typename U, std::enable_if_t<std::is_same<std::remove_cv_t<T>, std::remove_cv_t<U>>::value, int> = 0>
```
Comment (antonionevado) [inline dlk/python/dlk/templates/include/global.tpl.h]: Maybe we can rewrite this as:
```c++
template <typename U, std::enable_if_t<std::is_same<base_t, std::remove_cv_t<U>>::value, int> = 0>
```
to make it easier to read?

File: dlk/python/dlk/templates/manual/consts/input.tpl.cpp (hunk 7 above)
Comment (primenumber) [reply dlk/python/dlk/templates/manual/consts/input.tpl.cpp]: Add `static`.

File: dlk/python/dlk/core/data_types.py (hunk 9 above)
Comment (primenumber) [reply dlk/python/dlk/core/data_types.py]: Remove these types except `PackedUint64`.

File: dlk/python/dlk/templates/include/global.tpl.h (hunk 10 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/global.tpl.h]: Fixed.

**Comment by primenumber** [comment]
@n-nez @antonionevado Fixed these points. Plz re-review.

**Comment by n-nez** [comment]
@primenumber `PackedUint8` seems to be still here. Could you remove it as well? :thinking:

**Comment by primenumber** [comment]
@n-nez Removed `PackedUint8`.

**Comment by iizukak** [comment]
@n-nez Can you check this PR?

**Comment by n-nez** [comment]
Thank you! LGTM!

**Pull request merged**